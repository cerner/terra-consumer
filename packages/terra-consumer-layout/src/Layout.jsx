import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import Nav from 'terra-consumer-nav';
import ResponsiveElement from 'terra-responsive-element';
import Overlay from 'terra-overlay';
import { injectIntl, intlShape } from 'react-intl';
import 'terra-base/lib/baseStyles';
import styles from './Layout.scss';

const cx = classNames.bind(styles);

const propTypes = {
  /**
  * Object of configuration for the side navigation and profile.
  */
  nav: PropTypes.object.isRequired,
  /**
  * Array of links to show for the content of the help button
  */
  helpItems: PropTypes.array,
  /**

  * Alert banner
  */
  siteAlert: PropTypes.element,

   /* A center justified logo in header for mobile.
   */
  logo: PropTypes.shape({
    /**
     * A center justified logo in header for mobile.
     */
    mobileLogo: PropTypes.shape({
      url: PropTypes.string.isRequired,
      altText: PropTypes.string.isRequired,
    }),
    navLogo: PropTypes.object,
  }),

  /**
   * Injected react-intl formatting api
   */
  intl: intlShape.isRequired,
  children: PropTypes.node.isRequired,
};

class Layout extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      isMobileNavOpen: false,
    };

    this.toggleNav = this.toggleNav.bind(this);
  }

  toggleNav() {
    this.setState({
      isMobileNavOpen: !this.state.isMobileNavOpen,
    });
  }
  render() {
    const { nav, helpItems, logo, siteAlert, intl, ...customProps } = this.props;
    const overlay = (
      <Overlay
        onRequestClose={this.toggleNav}
        isOpen={this.state.isMobileNavOpen}
        backgroundStyle="clear"
        isRelativeToContainer
      />
    );

    let navConfig = nav;
    if (logo && logo.navLogo) {
      navConfig = Object.assign({}, nav, { logo: logo.navLogo });
    }

    return (
      <div className={cx('wrap')}>
        <div className={cx('skip-container')}>
          <a className={cx('skip-to-main-content')} href="#main-container" id="skip-maincontent-link">
            {intl.formatMessage({ id: 'Terra.Consumer.Layout.skipcontent' })}
          </a>
        </div>
        {siteAlert}
        <div {...customProps} className={cx('layout', { open: this.state.isMobileNavOpen }, customProps.className)}>
          <nav className={cx('nav')}>
            <Nav
              {...navConfig}
              onRequestClose={this.toggleNav}
            />
          </nav>
          <main id="main-container" className={cx('main-container')}>
            <ResponsiveElement defaultElement={overlay} responsiveTo="window" medium={<div />} />
            <div className={cx('main-container-inner')}>
              <div className={cx('nav-burgerbar')}>
                <Nav.Burger handleClick={this.toggleNav} />
                <div className={cx('mobile-logo')}>
                  {logo && logo.mobileLogo &&
                    <img src={logo.mobileLogo.url} alt={logo.mobileLogo.altText} />
                  }
                </div>
              </div>
              <div className={cx('main-content')}>{this.props.children}</div>
              <Nav.Help className={cx('help-button')} helpNavs={helpItems} id="nav-help-button" />
            </div>
          </main>
        </div>
      </div>
    );
  }
}

Layout.propTypes = propTypes;

export default injectIntl(Layout);
