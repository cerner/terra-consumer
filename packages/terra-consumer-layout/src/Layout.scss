@import './variables';

:root {
  // Body
  --terra-consumer-body-background-image: url('https://s3.amazonaws.com/healthelastic-prod/terra-consumer-background-images.jpg');
  --terra-consumer-body-background-color: #c7d4ea;

  // Nav
  --terra-consumer-nav-background-color: rgba(255, 255, 255, 0.25);
  --terra-consumer-nav-width: 320px;
  --terra-consumer-nav-link-color: #3d3d3d;

  // profile
  --terra-consumer-profile-link-border-background: var(--terra-consumer-body-background-image);
  --terra-consumer-profile-link-background: var(--terra-consumer-nav-background-color);
  --terra-consumer-profile-link-width: var(--terra-consumer-nav-width);
  --terra-consumer-profile-link-color: var(--terra-consumer-nav-link-color);
}


:local {
  .wrap {
    background-attachment: fixed;
    background-color: var(--terra-consumer-body-background-color);
    background-image: var(--terra-consumer-body-background-image);
    background-size: cover;
    overflow-x: hidden;

    @media screen and (max-width: $terra-medium-breakpoint) {
      background-image: none;
    }
  }

  .layout {
    display: flex;
    font-family: 'Roboto', sans-serif;
    -moz-osx-font-smoothing: grayscale;
    -webkit-font-smoothing: antialiased;
    margin: 0;
    min-height: 100vh;
    overflow: hidden;
  }

  .nav {
    min-width: var(--terra-consumer-nav-width);
    position: relative;
    width: var(--terra-consumer-nav-width);
    z-index: 100;

    @media screen and (max-width: $terra-medium-breakpoint) {
      background-color: var(--terra-consumer-body-background-color);
      box-shadow: 0 1px 3px rgba(28, 31, 33, 0.35);
      float: left;
      margin-left: calc(var(--terra-consumer-nav-width) * -1);
      transform: translate3d(0, 0, 0);
      transition: transform 0.35s ease;
      will-change: transform;
    }

    @media screen and (max-width: $terra-tiny-breakpoint) {
      margin-left: -100%;
      min-width: 0;
      width: 100%;
    }

    &::before {
      background-color: var(--terra-consumer-nav-background-color);
      bottom: 0;
      content: '';
      left: 0;
      position: absolute;
      right: 0;
      top: 0;
      z-index: -1;
    }
  }

  .open nav {
    @media screen and (max-width: $terra-medium-breakpoint) {
      transform: translate3d(var(--terra-consumer-nav-width), 0, 0);
      transition-duration: 0.5s;
    }

    @media screen and (max-width: $terra-tiny-breakpoint) {
      transform: translate3d(100%, 0, 0);
    }
  }

  .main-container {
    flex-grow: 1;
    max-width: 100%;
    text-align: left;
    white-space: break-word;

    .main-container-inner {
      padding: var(--terra-consumer-layout-padding, 10px);
    }
  }


  .open .main-container {
    @media screen and (max-width: $terra-medium-breakpoint) {
      overflow: hidden;
      position: fixed;
    }
  }

  .main-content {
    @media screen and (max-width: $terra-medium-breakpoint) {
      /* Set a min height for the content to keep the help button sticky to the
       * bottom of the screen when the content does not fill the viewport.
       * Help button with margin has height of 80 px. Height of help button and
       * some extra space subtracted from viewport height to stop scrolling in
       * empty layout.
       */
      min-height: calc(100vh - 120px);
    }
  }

  .nav-burger {
    display: block;
    font-size: 20px;
    left: 15px;
    padding: 0;
    position: absolute;
    top: 10px;
    z-index: 1;

    @media screen and (min-width: $terra-medium-breakpoint) {
      display: none;
    }
  }

  .help-button {
    margin: 15px;
    text-align: right;

    @media screen and (min-width: $terra-medium-breakpoint) {
      bottom: 15px;
      margin: 0;
      position: fixed;
      right: 15px;
    }
  }

  .skip-to-main-content {
    height: 1px;
    left: -1000px;
    overflow: hidden;
    position: absolute;
    top: auto;
    width: 1px;

    &:focus {
      height: auto;
      position: static;
      width: auto;
    }
  }

  .skip-container {
    text-align: center;
  }
}
