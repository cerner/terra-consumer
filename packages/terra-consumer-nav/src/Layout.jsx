import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import Nav from './Nav';
import styles from './Layout.scss';

const cx = classNames.bind(styles);

const propTypes = {
  // expand this
  nav: PropTypes.object,
};

const Layout = ({
  nav,
  ...customProps
}) => (
  <div className={cx('layout', customProps.className)}>
    <Nav {...nav} />
    <div className={cx('main-container')}>
      <h2>I am in the main content</h2>
    </div>
  </div>
);

Layout.propTypes = propTypes;

export default Layout;
