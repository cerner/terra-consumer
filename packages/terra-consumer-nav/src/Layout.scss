@import './variables';

:local {
  .layout {
    background: linear-gradient(#eee, #999);
    display: flex;
    font-size: 14px;
    height: 100vh;
    left: 0;
    max-width: 100vw;
    overflow: hidden;
    position: absolute;
    top: 0;
    width: 100vw;
  }

  .main-container {
    height: 100vh;
    left: $terra-consumer--nav-container-width;
    max-width: calc(100vw - #{$terra-consumer--nav-container-width});
    overflow: scroll;
    position: absolute;
    text-align: left;
    top: 0;
    width: calc(100vw - #{$terra-consumer--nav-container-width});
    @media screen and (max-width: $terra-medium-breakpoint) {
      left: 0;
      max-width: 100vw;
      width: 100vw;
    }
  }
}
