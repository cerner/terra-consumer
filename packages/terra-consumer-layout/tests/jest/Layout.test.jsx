import React from 'react';
import Layout from '../../src/Layout';

describe('Layout', () => {
  const defaultRender = <Layout />;

  // Snapshot Tests
  it('should render a default component', () => {
    const wrapper = shallow(defaultRender);
    expect(wrapper).toMatchSnapshot();
  });

  // Prop Tests
  it('should use the default value when no value is given', () => {
    const wrapper = shallow(defaultRender);
    expect(wrapper.find('.layout').text()).toEqual('defualt');
  });

  // Structure Tests
  it('should have the class layout', () => {
    const wrapper = shallow(defaultRender);
    expect(wrapper.prop('className')).toContain('layout');
  });
});
