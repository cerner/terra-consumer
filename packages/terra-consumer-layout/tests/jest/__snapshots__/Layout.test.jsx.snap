// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout should render a default component 1`] = `
<IntlProvider
  locale="en-US"
  messages={
    Object {
      "Terra.Consumer.Layout.skipcontent": "Skip to Main Content",
    }
  }
>
  <div>
    <InjectIntl(Layout) />
  </div>
</IntlProvider>
`;
